### PR TITLE
Non-default-deny network policies

### PR DESCRIPTION
**Pull request by squeed**
See contents.

Sample implementation [here](https://github.com/cilium/cilium/pull/30572).

### CODE REVIEW

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (joestringer) [inline cilium/CFP-30572-non-default-deny-policies.md:60]: I found myself falling into a bit of a trap with the wording around `non-matching`. In the newer ANP terminology they clearly define the _subject_ of the policy and separately the rules that apply to that subject. In this case, the "non-matching traffic" is all traffic that is part of the direction in the rules but not explicitly described in the rules, but also *still part of the subject*. Ie if you use an endpointSelector to apply policy for `app: foo` within namespace `foo`, then this has no impact on the policy for `app: bar` in namespace `bar`. The non-matching traffic is the traffic to/from the subject which is not explicitly described in the policy.

I think that this existing description is probably clear enough, but we might consider mentioning the subject:
```suggestion
Note that this is unrelated to the `IngressDeny` and `EgressDeny` policy expressions. Those dictate the action for *matching* traffic for the endpoint subject to the policy, whereas the `DefaultAction` specifies what should happen with *non-matching* traffic for the subject endpoint.
```

(Not entirely sure if this helps but wording is always a bit tough. Feel free to adopt, shuffle around this suggestion or reject it altogether)

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (joestringer) [inline cilium/CFP-30572-non-default-deny-policies.md:53]: My first impression was that expressing this feature as a `DefaultAction` creates some dissonance with the current language we use in CNP/CCNP. Specifically, I think of the current policy expressions in the CNP/CCNP as _user intent_ and not _orders for a thing to do something_. We tend to avoid using any sort of verbs or actions. The nice property of this is that we end up with a bunch of composable statements that have no reference to a specific time or place. When I read a policy, I read passive statements like:
- "when `app: foo` sends traffic to endpoints with `app: bar` on port 80, it is allowed"
- "when `app: foo` sends HTTP requests to `/public` on port 80 to `app: bar`, it is allowed".

Then, if we need to evaluate the way that each of these policies applies, we can tell individually by looking at one rule or another that the specific target traffic should be allowed. There's no sense of whether rules in one policy apply before or after rules in another policy. I think ultimately this is how we end up being able to process these expressions of user intent efficiently down into BPF for fast lookups, because there is no inherent ordering in the policy language. Of course, in the case of deny policies there is a slight deviation from this, and in the implementation details of L7 policies or so on we do end up needing to evaluate the semantics of the rules with respect to each other, but we're talking about a relatively small set of guidelines around how sets of two rules can be combined. It's not a general purpose multi-level rule priority system.

With that in mind I thought it's worth at least throwing around some ideas that could allow this feature to be expressed as a property of the policy itself without introducing verb language:

- `mode: PassByDefault/DenyByDefault`: Models the policy as having a mode of operation which is more like a property and less like a specific action at a time and place.
- `defaultAllow: true`: This is a bit specific rather than having a mode string, but has the benefit of being false by default without specific code to handle it, and I think that it clearly composes with the individual rules in the policy in a similar way to the current proposal.
- `nonMatchingTraffic: Pass/Drop`: A bit more explicit, this spells out what "default" means in the previous proposals.
- `zeroTrust: true`: assume a zero trust posture, ie do not allow anything unless it is explicitly allowed. Cons: potential for this to have overloaded meaning given the marketing origins of the expression "zero trust". This is also pretty opinionated, it implies anyone using this requested policy feature is not following zero trust. Then again, is that so far from reality....? If every user of the cluster declared their policy intent ahead of time, this feature would not be necessary at all I think?

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (joestringer) [inline cilium/CFP-30572-non-default-deny-policies.md:58]: One thing I like about the idea of a string for this property is the possibility for us to reuse this same field in future for a per-policy / per-namespace audit mode feature. May be worth considering whether we want to leave that option open here with a new enum value for the field or whether we would expect such a selective audit mode feature to be implemented separately.

**Review by joestringer — COMMENTED** [review]
Overall the feature here makes sense, my review below is mostly quibbling about naming/semantics and ensuring the wording is precise in order to minimize confusion as much as possible.

I don't anticipate many concerns in general with the proposal here, happy to approve once we're on the same page for semantics. Obviously when it comes to implementation details we'll need to think through how the docs can express this feature in clear terms, and also the sorts of test cases we will have to validate the feature. The four precedence examples already provide some hints as to where some of the testing effort should be spent.

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (joestringer) [reply cilium/CFP-30572-non-default-deny-policies.md:53]: I think there are elements of the same concern in @learnitall's feedback over [here](https://github.com/cilium/cilium/pull/30572#pullrequestreview-1867389946), where a "defaultAction" could be interpreted as a "slot" in the overall architecture which is populated by this policy, rather than a property of this specific policy which then implies that it doesn't impact any other policy in any way.

Comment (joestringer) [reply cilium/CFP-30572-non-default-deny-policies.md:53]: Whichever pattern we choose, I think it's also worth reading through the UX of this feature / field and express it in human terms the way that @learnitall did at the bottom of his linked comment above. If it's natural to go from the field name to a human-friendly description of the behaviour then that will serve well for documentation and will help users to read and understand the policy when they see this field.

Comment (squeed) [reply cilium/CFP-30572-non-default-deny-policies.md:53]: I agree that the naming is awkward -- hence this CFP :-). We're trying to find a way to express - arguably - the most complicated part of k8s / Cilium network policy. Namely, that creating the first policy for a subject is really doing *two* things -- allowing X, and blocking ^X. I don't have any good answers :-).

Coming at this backwards, from the perspective of the developer, `nonMatchingTraffic: Pass/Drop` seems the closest fit. "Pass" as a term already has a meaning in the context of firewall rules, and it pretty closely matches what this PR does. The only awkwardness is that it - from a certain reading - could imply that all other traffic would be dropped, even if allowed by other network policies. Hm.

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (squeed) [reply cilium/CFP-30572-non-default-deny-policies.md:70]: Oh, good point. We use "Rule" and "Policy" somewhat interchangeably in the codebase, since a `CiliumNetworkPolicy.Spec` is just a `struct Rule`. And then `Rule.Ingress []IngressRule` further complicates matters.

This should be a policy-level field. I'll update the doc.

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (squeed) [reply cilium/CFP-30572-non-default-deny-policies.md:126]: Agreed, once I sat down to think about agent config flags (which had been proposed somewhere), I realized I preferred API types over config.

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (squeed) [reply cilium/CFP-30572-non-default-deny-policies.md:53]: So [upstream k8s](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.29/#networkpolicyspec-v1-networking-k8s-io) has the `policyTypes` array, which solves a similar problem as this one. That said, I don't like the connotations implied by "Type", which seems wrong for this kind of field.

Perhaps
```yaml
spec:
  ...
  nonMatchingTraffic:
    ingress: Drop
    egress: Pass
```

is the right shape. We can compute the defaults neatly, based on the presence or absence of rules in each direction.

Comment (joestringer) [reply cilium/CFP-30572-non-default-deny-policies.md:53]: Incidentally the one thing I find a little ambiguous about the `nonMatchingTraffic` proposal is exactly what the "non-matching" aspect relates to. If you just look at the hierarchy proposed above with `spec: nonMatchingTraffic: ...`, the field is nested directly under the specification. As such it's not immediately clear the relationshp between `nonMatchingTraffic` and the `endpointSelector` since they're sibling fields. I think we all mean that `nonMatchingTraffic` still must match the `endpointSelector`. A couple of thoughts further on that would be either to embed that understanding directly into the naming ("subjectDefaultPosture: Pass/Drop"?), to move this field to become a property of the `EndpointSelector`, or to express this feature as a rule under the `ingress`/`egress` stanzas (not exactly great, because then it's just one more rule in a list of rules and it's much easier to lose this highly important policy-wide semantic in a big list of rules).

I actually think that the use of `default` wording provides just enough abstraction to cover this ambiguity. Default relies on the user understanding the notion that "a policy creates a default deny posture", but once they have that knowledge they can quickly jump from a field named something like `PassByDefault` to an intuitive understanding of how that changes the policy behaviour.

Since these options are probably fractal, maybe to put forth what seems to me one of the better options mitigating most of the downsides discussed above, how about:

```
spec:
  endpointSelector:
    matchLabels:
      org: empire
      class: deathstar
    defaultPosture: Pass
```

Comment (joestringer) [reply cilium/CFP-30572-non-default-deny-policies.md:53]:
```
  nonMatchingTraffic:
    ingress: Drop
    egress: Pass
```

Just to expand a bit more specifically on explicitly defining the `ingress` / `egress` aspects here. It is more explicit, which can make it clearer to reason about. To explore the semantics a bit we can probably iterate through various permutations of policies with the fields specified or unspecified and the implied meaning.

## Policy behaviour today

```
spec:
  endpointSelector:
    matchLabels:
      org: empire
      class: deathstar
  ingress:
    ...
```

Ingress rules are specified. The subject "deathstar" endpoints are put into default deny at ingress. Some specific traffic is also allowed, as specified in the rules. Other policies may further extend the allowed/denied traffic at ingress. This policy does not impact egress.

---

```
spec:
  endpointSelector:
    matchLabels:
      org: empire
      class: deathstar
  ingressDeny:
    ...
```

IngressDeny rules are specified. The subject "deathstar" endpoints are put into default deny at ingress. Some specific traffic is also denied, as specified in the rules. Other policies may further extend the allowed/denied traffic at ingress. No other policy may allow traffic that is specified in the deny rules here. This policy does not impact egress.

---

```
spec:
  endpointSelector:
    matchLabels:
      org: empire
      class: deathstar
  ingress:
  - {}
```

Ingress rules are specified, and there is an ingress rule which specifies no aspects of traffic to match. Due to the existance of the `ingress` stanza, the endpoint goes into default deny, and furthermore since the only rule does not match any traffic, this implicitly denies all traffic. Other policies may further extend the allowed/denied traffic at ingress. This policy does not impact egress.

## Split defaults proposal

I'll use the `nonMatchingTraffic` expression of this CFP idea to reason about the ingress/egress aspect below, but I think this will equally apply if we choose to express the feature in other terms.

```
spec:
  endpointSelector:
    matchLabels:
      org: empire
      class: deathstar
  ingress:
    ...
```

Like today, we must not change the semantics of this. Due to the lack of `nonMatchingTraffic` field and the fact that `ingress` is specified, the "deathstar" endpoints are put into default deny at ingress. This policy does not impact egress. Implicitly this is treated as though the above policy also has the following `nonMatchingTraffic` specification:

```
  nonMatchingTraffic:
    ingress: Drop
    egress: Pass
```

We implicitly construct this nonMatchingTraffic behaviour based on the existence of the `ingress` field and the lack of existance of the `egress` field.

---

```
spec:
  endpointSelector:
    matchLabels:
      org: empire
      class: deathstar
  nonMatchingTraffic:
    ingress: Pass
  ingress:
    ...
```

Now that the `nonMatchingTraffic` field explicitly specifies the `ingress` aspect, the `ingress` field no longer implicitly creates a default deny posture. Of course given that there is no explicit mention of egress, this policy has no impact on egress.

---

```
spec:
  endpointSelector:
    matchLabels:
      org: empire
      class: deathstar
  nonMatchingTraffic:
    egress: Deny
  ingress:
    ...
```

There is no `nonMatchingTraffic` specification on ingress, so given that there are ingress rules the default posture on ingress is deny. The `nonMatchingTraffic` field specifies the posture for egress though, so now egress is set to a default deny (which may be overridden by other policies).

---

```
spec:
  endpointSelector:
    matchLabels:
      org: empire
      class: deathstar
  nonMatchingTraffic:
    ingress: Pass
  ingress:
  - {}
```

Although there is a rule in the `ingress` stanza, that rule doesn't select any specific traffic, because the entire rule is empty. That implies that all traffic on ingress to this endpoint is non-matching and hence subject to the `nonMatchingTraffic` behaviour.

What's interesting about this case is that if we dropped/ignored the `nonMatchingTraffic` field, this would be the typical expression of an [ingress default deny policy](https://docs.cilium.io/en/stable/security/policy/language/#ingress-egress-default-deny). 

Strictly speaking I think this is actually consistent and I think that if someone was to read through such a policy, given the clear `nonMatchingTraffic` field I think the behaviour could be intuitively understood.

---

Based on the examples above, there will still need to be default behaviour for the cases where the user does not explicitly specify the new field. However, the benefit would be that it's possible to express the variations on default behaviour explicitly as fields directly in the policy. We could even provide tooling for existing users to convert their underspecified policies directly into explicitly specified policies, ie take existing policies and populate the new field to make the policies clearer. I don't feel strongly about this either way but I think that the explicit split of the directions here is a reasonable proposal.

Comment (joestringer) [reply cilium/CFP-30572-non-default-deny-policies.md:53]: Recording back to this thread: @nathanjsweet , @learnitall , @squeed and I discussed the "split defaults" aspect of this for splitting ingress/egress specifications for the behaviour and generally agreed that it's better to explicitly split this feature by direction.

We discussed through some of the variations on naming that I brought up at the start of this thread and discussed nuances on semantics. We didn't conclusively come to an agreement on that but I think we're more in sync about the concerns so hopefully we can drive agreement soon.

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (joestringer) [reply cilium/CFP-30572-non-default-deny-policies.md:58]: From discussion out-of-band with @nathanjsweet , @learnitall , @squeed we acknowledged the potential overlap for describing audit mode in this way, but also identified that there are nuances in the way that audit mode applies which raise some uncertainty about whether embedding audit mode into individual policies is a good idea. Specifically, audit mode is typically applying on a per-endpoint basis rather than selecting specific traffic to apply to. While it's plausible for us to express this concept in the policy in this area, we don't want to tie the overall proposal here to the possibility of audit mode, so it makes sense to put this idea back on the shelf for now. If and when we have stronger motivation to address a more dynamic audit mode feature for individual endpoints/namespaces or even policies, we can revisit the solution space specifically for that feature.

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (squeed) [reply cilium/CFP-30572-non-default-deny-policies.md:60]: I changed the wording a bit, using the terms "subject" and "peer". This is at least vaguely consistent with the k8s terminology.

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (squeed) [reply cilium/CFP-30572-non-default-deny-policies.md:77]: Right, the documentation is a bit confusing as well, since it mentions "deny policies". Deny is not a property of a policy, it's a property of a rule. The go types are, loosely,
```go
type PolicySpec struct {
  Ingress []IngressRule
  IngressDeny []IngressRule
  //...
}
```

I'll give the CFP a once-over to make it consistent, but from the perspective of implementation, a Policy consists of a set of Rules, and Rules have a verdict. Kind of.

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (joestringer) [reply cilium/CFP-30572-non-default-deny-policies.md]: The "vetoing" property is the critical semantic here, I agree. If I visualize the network policy engine as having a map of rules and then a "slot" at the end of processing which decides whether or not to drop all traffic that didn't already receive a verdict, then I would say that today each network policy populates that slot. This field is supposed to decide whether or not this specific policy will populate that slot. As long as any one policy populates that slot, there will be a default deny. 

I don't love that omitting the field is the same as setting it to true, but I don't see a great way to reverse the expression of the feature and retain the veto property. Certainly not `defaultAllow`, because that would imply that this policy may override other policies. Similarly for `defaultPass`. Although I had some fondness for the `PassByDefault`, `DenyByDefault` proposals I made above, I do not think that this feature can be expressed as a "mode" in the policy. There is a subtlety here in the way that each individual policy interacts with one another and with the core policy engine to potentially overlap with static agent flags. I guess that's the price of fixing API mistakes from years ago :)

Comment (learnitall) [reply cilium/CFP-30572-non-default-deny-policies.md]: I think that's a good call @squeed. It's very clear and there probably isn't a great way to express this idea without using a verb. I don't know if there is a great way to make this field false by default either, since reversing the logic would add some bloat:

```yaml
spec:
  doesNotCreateDefaultDeny:
    ingress: false
    egress: false
```

The last bit of feedback I have here is that there might be a synonym for "creates" that sounds less verb-y but isn't necessarily in the form of third person singular:

* Establish
* Enact
* Initiate
* Trigger (my favorite)

Comment (joestringer) [reply cilium/CFP-30572-non-default-deny-policies.md]: I agree `doesNot...` is too clumsy.

Maybe one more for the list, `EnableDefaultDeny`. That would probably be my pick out of all of the verbs above including `Create`.

Initiate/Trigger seem too temporal to me, they suggest an even more timing-sensitive aspect to the semantic which I don't think reflects the reality of the feature. Enact is a bit more nuanced, perhaps less accessible for non-English speakers. Establish is fine. Create is OK, it fits CRUD terminology better, not sure if that's a good thing or bad thing since we don't have the rest of the CRUD verbs.

Comment (joestringer) [reply cilium/CFP-30572-non-default-deny-policies.md]: Maybe one other subtle nuance to throw into the ring, replace `Default` with `Implicit`. This wording comes directly from one of the users that was requesting the feature. So something like `CreateImplicitDeny` / `EnableImplicitDeny`.

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (joestringer) [reply cilium/CFP-30572-non-default-deny-policies.md]: Cool yep. We can always improve these things over time. If we're at least on the same page for how this CFP treats these words and that we should try to use this terminology more going forward then things should become clearer :+1:

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (squeed) [reply cilium/CFP-30572-non-default-deny-policies.md]: I like "enable", seems easy enough. I do prefer "Default" over "Implicit", since firewalls have had the notion of a "default action" since "time immemorial".

I'm going to update the PR with `enableDefaultDeny:...` and then this should be ready to go.

Comment (nathanjsweet) [reply cilium/CFP-30572-non-default-deny-policies.md]: > I like "enable", seems easy enough. I do prefer "Default" over "Implicit", since firewalls have had the notion of a "default action" since "time immemorial".

So is the default value "true"?

Comment (squeed) [reply cilium/CFP-30572-non-default-deny-policies.md]: The default value is true for each direction with rules. Same as before -- creating a policy with just Ingress rules blocks all other ingress traffic, but leaves egress untouched.

Comment (nathanjsweet) [reply cilium/CFP-30572-non-default-deny-policies.md]: I agree with the overall semantics, but my preference is that default boolean values are always false. I know there are exceptions all over the place in the Cilium code base, but I would prefer `DefaultAllow` or `enableDefaultAllow` over `enableDefaultDeny` so that the default boolean value can be "false" as I think most API users expect it. However, I'm not going to hold up this CFP over syntax.

Comment (joestringer) [reply cilium/CFP-30572-non-default-deny-policies.md]: @nathanjsweet I sympathize with the argument around boolean values, but specifically if we look at `enableDefaultAllow`, it implies that this policy creates an "allow" rule by default, which does not compose as clearly with other policies. If you have three policies, two with `enableDefaultAllow: true` and one without, the behaviour may surprise users. The policy without the field will implicitly have `enableDefaultAllow: false`, and this results in the actual behaviour being to enable default deny. The (omitted) `false` boolean would be overriding the explicit positive from the other policies.

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (christarazi) [inline cilium/CFP-30572-non-default-deny-policies.md:20]: Very nit: 
```suggestion
The administrator may wish to create a Clusterwide Cilium Network Policy (CCNP)
```

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (christarazi) [inline cilium/CFP-30572-non-default-deny-policies.md:81]:
```suggestion
Note that this is unrelated to the `IngressDeny` and `EgressDeny` policy expressions. Those dictate the action for selected peer traffic, whereas the `EnableDefaultDeny` specifies what should happen with unselected peer traffic.
```

**Review by christarazi — APPROVED** [review]
Two very minor nits, LGTM

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (joestringer) [reply cilium/CFP-30572-non-default-deny-policies.md:20]: Cilium comes first in the acronym in the vast majority of matches in the codebase, I think the wording is correct as-is.

```
$ git grep ClusterwideCi | wc -l
2
$ git grep CiliumClusterwide | wc -l
500
```

**Comment by joestringer** [comment]
@squeed I believe that as a committer, you should be able to merge this PR. I'll leave it up to you to make any last minute wording/typo changes and do the honors.

File: cilium/CFP-30572-non-default-deny-policies.md
Comment (christarazi) [reply cilium/CFP-30572-non-default-deny-policies.md:20]: Oh, you're right, I didn't check before posting and I've always thought it was Clusterwide first and then the CNP part after, but I guess I had the reverse in my head all this time :)

**Pull request merged**
